test/vegeta/pacers.go

```diff
@@ -0,0 +1,234 @@
+		// Running behind, send next hit immediately.
+		return 0, false
+	}
+
+	// Re-arranging our hits equation to provide a duration given the number of
+	// requests sent is non-trivial, so we must solve for the duration numerically.
+	// math.Round() added here because we have to coerce to int64 nanoseconds
+	// at some point and it corrects a bunch of off-by-one problems.
+	nsPerHit := 1 / sup.hitsPerNs(elapsedTime)
+	hitsToWait := float64(elapsedHits+1) - expectedHits
+	nextHitIn := time.Duration(nsPerHit * hitsToWait)
+
+	// If we can't converge to an error of <1e-3 within 10 iterations, bail.
+	// This rarely even loops for any large Period if hitsToWait is small.
+	for i := 0; i < 10; i++ {
+		hitsAtGuess := sup.hits(elapsedTime + nextHitIn)
+		err := float64(elapsedHits+1) - hitsAtGuess
+		if math.Abs(err) < 1e-3 {
+			return nextHitIn, false
+		}
+		nextHitIn = time.Duration(float64(nextHitIn) / (hitsAtGuess - float64(elapsedHits)))
+	}
+
+	return nextHitIn, false
+}
+
+// hits returns the number of expected hits for this pacer during the given time.
+func (sup *steadyUpPacer) hits(t time.Duration) float64 {
+	// If t is smaller than the upDuration, calculate the hits as a trapezoid.
+	if t <= sup.upDuration {
+		curtHitsPerNs := sup.hitsPerNs(t)
+		return (curtHitsPerNs + sup.minHitsPerNs) / 2.0 * float64(t)
+	}
+
+	// If t is larger than the upDuration, calculate the hits as a trapezoid + a rectangle.
+	upHits := (sup.maxHitsPerNs + sup.minHitsPerNs) / 2.0 * float64(sup.upDuration)
+	steadyHits := sup.maxHitsPerNs * float64(t-sup.upDuration)
+	return upHits + steadyHits
+}
+
+// hitsPerNs returns the attack rate for this pacer at a given time.
+func (sup *steadyUpPacer) hitsPerNs(t time.Duration) float64 {
+	if t <= sup.upDuration {
+		return sup.minHitsPerNs + float64(t)*sup.slope
+	}
+
+	return sup.maxHitsPerNs
+}
+
+// hitsPerNs returns the attack rate this ConstantPacer represents, in
+// fractional hits per nanosecond.
+func hitsPerNs(cp vegeta.ConstantPacer) float64 {
+	return float64(cp.Freq) / float64(cp.Per)
+}
+
+// combinedPacer is a pacer that combines multiple pacers and runs them sequentially when being used for attack.
+type combinedPacer struct {
+	// pacers is a list of pacers that will be used sequentially for attack, must be non-empty.
+	pacers []vegeta.Pacer
+	// durations is the list of durations for the given Pacers, must have the same length as Pacers.
+	durations []time.Duration
+
+	// totalDuration is sum of the given Durations.
+	totalDuration uint64
+	// stepDurations is the accumulative duration of each step calculated from the given Durations.
+	stepDurations []uint64
+
+	curtPacerIndex  uint
+	prevElapsedHits uint64
+	prevElapsedTime uint64
+}
+
+// NewCombinedPacer returns a new CombinedPacer with the given config.
+func NewCombinedPacer(pacers []vegeta.Pacer, durations []time.Duration) vegeta.Pacer {
+	if len(pacers) == 0 || len(durations) == 0 || len(pacers) != len(durations) || len(pacers) == 1 {
```

Comment:
I think 1 is a valid combined pacer (redundant, yes), but seems valid.

Reply:
One reason for disallowing 1 is I failed to find a clean way to handle it :(
Since if only 1 pacer is provided, for a given `elapsedTime`, we'll need to know whether the pacer needs to be restarted. To determine it, we'll need to keep track of the `elapsedTime` for the last hit, and compare it with the current `elapsedTime` after `%duration`. It will add one extra state and make the code logic slightly more complicated.
I would prefer not to allow 1 pacer for now, but add it if it turns out to be a valid use case for us.